src/mentorship/ml/models/common.py

```diff
@@ -0,0 +1,39 @@
+from sklearn.base import clone
+
+
+class SplitPipeline:
+    def __init__(self, base_pipeline):
+        self.base_pipeline = base_pipeline
+        self.pipelines_ = {}
+
+    def fit(self, X, y):
+        for family in X['family'].unique():
+            indexer = X[X['family'] == family].index
+            X_part = X[X['family'] == family]
+            y_part = y.loc[indexer]
+
+            pipeline = clone(self.base_pipeline)
+            pipeline = pipeline.fit(X_part, y_part)
+
+            self.pipelines_[family] = pipeline
+        return self
+
+    def predict(self, X):
+        X.loc[:, 'forecast'] = 0
+        for family in X['family'].unique():
+            X_part = X[X['family'] == family]
+            X_part = X_part.drop('forecast', axis=1)
+            pipeline = self.pipelines_[family]
+            X.loc[X_part.index, 'forecast'] = pipeline.predict(X_part)
+
+        y_pred = X['forecast']
+        X = X.drop(columns='forecast')
+        return y_pred
```

Comment:
In order to get rid of temporary modification of `X` let's do the following:
```suggestion
    def predict(self, X):
        y_preds = []
        for key, X_part in X.groupby(self.split_key):
            pipeline = self.pipelines_[key]
            y_pred = pipeline.predict(X_part)
            y_preds.append(pd.Series(data=y_pred, index=X_part.index, name='forecast'))

        y_pred = pd.concat(y_preds).loc[X.index]
        return y_pred
```